### PR TITLE
test: migrate check_cfg to snapbox

### PR DESCRIPTION
**Pull request by eth3lbert**
### What does this PR try to resolve?

Part of https://github.com/rust-lang/cargo/issues/14039.

Migrate following to snapbox:

- `tests/testsuite/check_cfg.rs`

---

This commit allow deprecation for most of the tests currently in this file. This is because the tests either use `does_not_contain` or perform multiple `contains` assertion for single tool with macro `x!`.

Here are some potential options to further eliminate the deprecation:

  1. Remove the `x!` macro and replace the it with `str![]` for both Windows and non-Windows platforms.
  2. Modify the `x!` macro to only generate the `--check-cfg` part for different platforms and manually compose it with `format!`.

### CODE REVIEW

**Comment by eth3lbert** [comment]
Failed in CI with the following:
```
---- expected: tests/testsuite/lto.rs:372:13
++++ actual:   stderr
   1    1 | [COMPILING] foo v0.0.0 ([ROOT]/foo)
   2    2 | [RUNNING] `rustc --crate-name a [..]-C lto [..]`
   3    3 | [RUNNING] `rustc --crate-name b [..]-C lto [..]`
   4    4 | [RUNNING] `rustc --crate-name foo [..]-C lto [..]--test [..]`
   5    5 | [RUNNING] `rustc --crate-name foo [..]--crate-type bin [..]-C lto [..]`
   6      - [FINISHED] `release` profile [optimized] target(s) in [ELAPSED]s
   7    6 | [RUNNING] `[ROOT]/foo/target/release/deps/foo-[HASH][EXE]`
   8    7 | [RUNNING] `[ROOT]/foo/target/release/deps/a-[HASH][EXE]`
   9    8 | [RUNNING] `[ROOT]/foo/target/release/deps/b-[HASH][EXE]`
        9 + [FINISHED] `release` profile [optimized] target(s) in 1m 00s
```

This should be resolved once #14233 is merged.

**Comment by epage** [comment]
#14233 is now merged.

File: tests/testsuite/check_cfg.rs [hunk 1]
```diff
@@ -256,6 +269,7 @@ fn features_test() {
         .run();
 }
 
+#[allow(deprecated)]
 #[cargo_test]
 fn features_doctest() {
```
Comment (epage) [inline tests/testsuite/check_cfg.rs:274]: I think the `does_not_contains` could be removed.  If we were passing `-Z`, it would fail as this test runs on stable

**Comment by epage** [comment]
> Remove the x! macro and replace the it with str![] for both Windows and non-Windows platforms.

Overall, not a fan of `x!`.

**Comment by epage** [comment]
To add though, we can merge this now-ish (maybe remove some of the weirder `does_not_contains`) and further improve it in follow up PRs, rather than do it all at once.

**Comment by eth3lbert** [comment]
> To add though, we can merge this now-ish (maybe remove some of the weirder `does_not_contains`) and further improve it in follow up PRs, rather than do it all at once.

I remove all the `-Z`. However, if we only want to remove the one within the `features_doctest` function, please let me know!

**Comment by Urgau** [comment]
The `x!` macro was added very early the development of the feature in Cargo and has always been awkward, it was meant as a simple way to create the cmd argument without relying to much on the actual argument; in part because the syntax was still evolving and changing. It also was a convenient way to handle the Windows vs non-Windows escaping.

It has serve us well, but since then many things have evolved, `--check-cfg` is now stable, Cargo has a new testing engine...

In other words, feel free to remove the macro.

**Comment by eth3lbert** [comment]
> The `x!` macro was added very early the development of the feature in Cargo and has always been awkward, it was meant as a simple way to create the cmd argument without relying to much on the actual argument; in part because the syntax was still evolving and changing. It also was a convenient way to handle the Windows vs non-Windows escaping.
> 
> It has serve us well, but since then many things have evolved, `--check-cfg` is now stable, Cargo has a new testing engine...
> 
> In other words, feel free to remove the macro.

Thanks for the explanation and the addition from the author's perspective.

**Comment by epage** [comment]
I'll merge as-is and we can handle `x!` in a follow up

@bors r+
